contracts/src/PDPService.sol

```diff
@@ -181,6 +212,8 @@ contract PDPService {
         uint256 rawSize;
     }
 
+    // Appends new roots to the collection managed by a proof set.
+    // These roots won't be selected for proving until the next proving period.
```

Comment:
Lets think about this thoroughly.  There is state asymmetry
### State that `add` updates eagerly and `scheduleRemove` updates lazily
 - sumTree
 - leafCounts
 - leafCount
 - nextRootID

### State both update lazily
- lastChallengedLeaf

### Abstract asymmetry
Both new add and scheduleRemove are invisible to provePossession in terms of correctness and visible by performance though add is much more perf visible via sumtree.  That's probably the biggest abstract asymmetry I see but it only shows up through gas cost.

New adds are visible to other adds.  We need this for eager sequencing which is necessary.   New scheduleRemovals are not visible to other adds but adds never depended on any previous removes because proofsets are append only.  So there's no new symmetry break.

New adds are visible to scheduleRemoves in that when scheduled removals are processed they can remove new adds that have not yet been put into the proving set.  This is a better way to do things imo than to require all new adds are put in the proof set for one proving period before they can be deleted. Removes don't need to consider other removes because remove operation is idempotent.

Reply:
State asymmetry in sumtree, leafcounts and leafCount could be replaced by enqueuing add operations for processing in nextProvingPeriod like removals.  This would also get rid of the big gas cost asymmetry since sumtree find cost only changes each proving period. 

nextRootID state asymmetry exists to support good add sequencing.  If we enqueued adds we could instead update this only in nextProvingPeriod and check the length of the queue to return correct sequence numbers.

Reply:
I think the trade off is asymmetry in order to do adds eagerly.  I prefer eager operations for user flexibility unless its impossible to avoid them, as in the case of removals.  So I think the asymmetry is worth it.

Reply:
Well one pareto improvement is maybe to keep a variable for uncommitted leaf count and then add this to leafCounts at the end of nextProvingPeriod.  

Then we get rid of "lastChallengedLeaf" and prove against leafCount.

Reply:
No I actually don't like this because the sumtree needs a consistent view of the total leaves in order to work well so we'd need to add the uncommitted leaf count to total leafCount each time we ran a sumtree operation which seems worse.